rclcpp/include/rclcpp/sched_base.hpp

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include <sched.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <pthread.h>
+#include <atomic>
+
+
+namespace rclcpp {
+// forward declarations
+namespace executors {
+    class SingleThreadedExecutor;
+}; // rclcpp::executors
+namespace sched {
+
+struct pthread_struct {
+    void* __unused[90]; // Hacking here, may be platform dependent.
+    pid_t tid;
+};
+
+struct SchedAttr {
+    uint32_t size = sizeof(SchedAttr);              /* Size of this structure */
+    uint32_t sched_policy = SCHED_OTHER;      /* Policy (SCHED_*) */
+    uint64_t sched_flags;       /* Flags */
+    int32_t sched_nice;        /* Nice value (SCHED_OTHER, SCHED_BATCH) */
+    uint32_t sched_priority;    /* Static priority (SCHED_FIFO, SCHED_RR) */
+
+    /* For SCHED_DEADLINE */
+    uint64_t sched_runtime;
+    uint64_t sched_deadline;
+    uint64_t sched_period;
+
+    /* Utilization hints, unused for our purpose,
+       may enable in the future*/
+    // uint32_t sched_util_min;
+    // uint32_t sched_util_max;
+};
+
+/** Since pthread does not expose pid to us, this is a hack to get the (linux) pid.
+ *  This might be dangerous and non-portable.
+ */
+inline pid_t
+get_pid(pthread_t threadid) {
+    auto pthread_id = ((pthread_struct*) threadid);
+    /* this may occur if the thread is detached from the current thread, use this
+       function before calling detach() */
+    if (pthread_id == nullptr) {
+        printf("nullptr is passed to get_pid!\n");
+        return 0;
+    }
+    return pthread_id->tid;
+}
+
+/** Equality for SchedAttr, use it to prevent unnecessary syscall. */
+inline bool
+operator==(const SchedAttr& lhs, const SchedAttr& rhs) {
+    if (&lhs == &rhs) {
+        return true;
+    }
+    uint64_t* a = (uint64_t*) &lhs;
+    uint64_t* b = (uint64_t*) &rhs;
+    size_t offset = 0;
+    while (offset++ < (sizeof(SchedAttr) / 8)) {
+        if (*(a++) != *(b++)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+/** Inequality for SchedAttr, use it to prevent unnecessary syscall. */
+inline bool
+operator!=(const SchedAttr& lhs, const SchedAttr& rhs) {
+    return !(lhs==rhs);
+}
+
+inline long
+syscall_sched_setattr(pid_t pid, SchedAttr* sched_attr) {
```

Comment:
I think you can declare the sched_attr as const.

Reply:
Addressed.